# 00 · Preflight: Python and NumPy from the beginning

This chapter is a gentle introduction to the programming tools used in the rest
of the course. It assumes that you may be opening a Python notebook for the
first time. No knowledge of GeoDef, geodesy, or linear algebra is required.

We will begin with running a single notebook cell and naming a number. From
there, we will build toward numerical arrays, plots, and the matrix calculations
that appear throughout scientific computing. The examples use familiar Earth
science quantities such as distance, depth, elevation, and temperature, but the
programming ideas are general.

**Learning objectives**

By the end of the chapter, you will be able to:

- run and edit notebook cells without losing track of their order
- store numbers and text in variables and inspect their types
- use arithmetic, comparisons, lists, conditions, loops, and functions
- import NumPy and matplotlib and understand their common short names
- create, inspect, select from, and calculate with NumPy arrays
- make a clearly labeled scientific plot
- distinguish ordinary multiplication from matrix multiplication
- solve a small system of equations and a least-squares problem

**Prerequisites:** none  
**Estimated time:** 3–4 hours, including the exercises

## 1. Working in a Jupyter notebook

A notebook is a document made from **cells**. A text cell, like this one,
explains an idea using ordinary writing and Markdown formatting. A code cell
contains Python instructions. When you run a code cell, Python follows those
instructions and displays any output immediately below it.

Click the next cell and press **Shift–Enter**. You can also use the triangular
Run button in the toolbar. The number that appears beside the cell records when
it ran. That number is useful because a notebook remembers variables in the
order cells were run, which may differ from their order on the page.

In [1]:
print("Hello, Earth scientist!")

Hello, Earth scientist!


Python can also use a code cell as a calculator. If the last line is a value or
calculation, Jupyter displays it without requiring `print`. Try changing one
number and running the cell again. Editing and rerunning small examples is one
of the fastest ways to learn.

In [2]:
2 + 3

5

**Checkpoint.** If a cell shows an old output after you edit it, has the new
code run yet?

<details><summary>Answer</summary>

No. An output remains visible until the cell is run again. The safest check is
**Restart Kernel and Run All Cells**, which clears the notebook's memory and
runs every code cell from top to bottom.

</details>

## 2. Errors are information

Everyone who writes code encounters errors. An error message is Python's report
of where it became unable to continue. The last line usually gives the most
useful clue: for example, `NameError` means that Python does not know a name,
while `SyntaxError` often means that punctuation or spelling made an instruction
impossible to read.

The cell below is the corrected version of `prnt("Hello")`. If you want to
practice reading an error, temporarily remove the letter `i`, run the cell, read
the last line of the message, and then repair it.

In [3]:
print("The spelling is fixed.")

The spelling is fixed.


A **comment** begins with `#`. Python ignores everything after the `#` on that
line. Comments can record a fact that is not obvious from the calculation, such
as the physical meaning or units of a value. They should help a future reader,
not merely repeat the code in English.

In [4]:
depth_km = 12.5  # Depth below the surface, in kilometers
print(depth_km)

12.5


**A useful debugging routine:** read the final line of the error, find the
highlighted line of your code, check names and punctuation, then make one small
change and rerun the cell. Errors are normal evidence, not a sign that you are
unsuited to programming.

## 3. Variables: giving values useful names

A **variable** is a name that refers to a value stored in the notebook's memory.
The equals sign assigns the value on its right to the name on its left. It does
not mean “is mathematically equal forever.” A later cell can assign a new value
to the same name.

Clear names make scientific code readable. Names may contain letters, numbers,
and underscores, but cannot begin with a number. Python treats uppercase and
lowercase letters as different, so `depth_km` and `Depth_km` are different
names.

In [5]:
sample_name = "Sandia granite"
density_kg_m3 = 2670
print(sample_name, density_kg_m3)

Sandia granite 2670


A notebook keeps the latest assigned value. This is convenient, but it also
means that running cells out of order can create confusing results. In the next
pair of lines, the second assignment replaces the first one.

In [6]:
elevation_m = 1600
elevation_m = 1612
print(elevation_m)

1612


**Checkpoint.** What does Python remember after `temperature_c = 20` followed
by `temperature_c = 23`?

<details><summary>Answer</summary>

The name refers to the most recently assigned value, `23`. Assignment changes
what the variable refers to; it does not create an equation that Python solves.

</details>

## 4. Numbers, text, and other basic types

A **type** describes what kind of value Python is storing and therefore what
operations make sense. An `int` is a whole number, a `float` is a number stored
with a decimal representation, a `str` is text surrounded by quotation marks,
and a `bool` is either `True` or `False`.

Scientific measurements are usually floats even when their displayed value is
a whole number. Text such as a station name is a string and cannot be added to a
measurement as if both were numbers.

In [7]:
station_count = 12
velocity_mm_yr = 4.7
station_name = "ABQ1"
is_active = True

In [8]:
print(type(station_count), type(velocity_mm_yr))
print(type(station_name), type(is_active))

<class 'int'> <class 'float'>
<class 'str'> <class 'bool'>


Functions named `int`, `float`, and `str` can convert compatible values. Be
careful: converting a float to an integer removes the fractional part rather
than performing ordinary rounding. Use `round` when rounding is what you mean.

In [9]:
measured_depth = 7.8
print(int(measured_depth), round(measured_depth))

7 8


## 5. Arithmetic expressions and comparisons

An **expression** is a combination of values, variables, and operations that
Python evaluates to produce a result. The operators `+`, `-`, `*`, and `/` mean
addition, subtraction, multiplication, and division. A double asterisk, `**`,
means “raise to a power.” Parentheses make the intended order of calculation
explicit.

Python understands scientific notation: `3e6` means \(3 	imes 10^6\). The
letter `e` here means “times ten raised to,” not the mathematical constant.

In [10]:
distance_m = 3.2e3
travel_time_s = 8.0
speed_m_s = distance_m / travel_time_s
print(speed_m_s)

400.0


In [11]:
radius_km = 2.0
area_km2 = 3.14159 * radius_km**2
print(area_km2)

12.56636


A comparison produces a Boolean value. Use `==` to ask whether two values are
equal; a single `=` performs assignment. Other comparisons include `<`, `>`,
`<=`, `>=`, and `!=` (“not equal”).

In [12]:
magnitude = 5.4
print(magnitude >= 5.0)
print(magnitude == 6.0)

True
False


## 6. Strings and readable output

A **string** is a sequence of text characters enclosed by single or double
quotation marks. Strings are useful for station names, units, labels, and
messages. Square brackets select characters by position; Python counts
positions from zero, so the first character has index `0`.

In [13]:
station = "ALBU"
print(station[0], station[-1])
print(len(station))

A U
4


An **f-string** places variable values inside text. Put the letter `f` before
the opening quote and enclose each variable or expression in braces. The part
`:.1f` below requests one digit after the decimal point. This is only display
formatting; it does not change the stored number.

In [14]:
depth_km = 12.46
message = f"The earthquake depth is {depth_km:.1f} km."
print(message)

The earthquake depth is 12.5 km.


## 7. Lists, positions, and slices

A **list** stores an ordered sequence of values. Lists are written with square
brackets and commas. They are useful for a small collection of labels or
values, although NumPy arrays will soon become our main container for numerical
data.

An **index** is a position in the sequence. Python begins at zero. Negative
indices count backward, so `-1` selects the last item.

In [15]:
minerals = ["quartz", "feldspar", "mica", "calcite"]
print(minerals[0])
print(minerals[-1])

quartz
calcite


A **slice** selects several positions. In `values[start:stop]`, the starting
position is included and the stopping position is excluded. Excluding the stop
may feel unusual at first, but it means that `values[:3]` contains exactly three
items: positions 0, 1, and 2.

In [16]:
depths_km = [2.1, 4.5, 7.0, 9.8, 12.3]
print(depths_km[:3])
print(depths_km[2:])

[2.1, 4.5, 7.0]
[7.0, 9.8, 12.3]


In [17]:
depths_km.append(15.0)
print(len(depths_km), depths_km)

6 [2.1, 4.5, 7.0, 9.8, 12.3, 15.0]


## 8. Making decisions with `if`

A program can choose which instructions to run. An `if` statement checks a
Boolean condition. The indented lines below it run only when that condition is
true. Indentation is part of Python's grammar, so the four spaces are not merely
decoration.

Optional `elif` (“else if”) branches check another condition, and `else` handles
everything that did not match an earlier branch.

In [18]:
magnitude = 4.2
if magnitude >= 5.0:
    print("Review this event promptly.")
else:
    print("Continue routine review.")

Continue routine review.


Python checks branches from top to bottom and stops at the first true condition.
For that reason, place the most restrictive threshold first in the example
below.

In [19]:
rainfall_mm = 37
if rainfall_mm >= 50:
    category = "heavy"
elif rainfall_mm >= 20:
    category = "moderate"
else:
    category = "light"
print(category)

moderate


## 9. Repeating work with loops

A **loop** repeats an indented block of code. A `for` loop takes one item at a
time from a sequence. Here, `depth` is a temporary name: during each pass
through the loop, it refers to the next value in `depths_km`.

In [20]:
depths_km = [1.5, 3.0, 7.2]
for depth in depths_km:
    print(depth * 1000, "m")

1500.0 m
3000.0 m
7200.0 m


The built-in `range` function produces a sequence of integers. `range(4)` gives
0, 1, 2, and 3. The stop value is excluded, just as it is in a slice.

In [21]:
for index in range(4):
    print("Sample", index)

Sample 0
Sample 1
Sample 2
Sample 3


Loops are important for understanding program flow and for tasks that genuinely
happen one at a time. For large numerical calculations, NumPy can often apply
one operation to an entire array. We will compare these approaches in Section
14.

## 10. Reusing a calculation with functions

A **function** is a named block of code that performs a task. We define one with
`def`. Names inside the parentheses are **parameters**, meaning placeholders for
input values. The `return` statement sends a result back to the code that called
the function.

Functions help us avoid copying a calculation and make its purpose explicit.

In [22]:
def kilometers_to_meters(distance_km):
    distance_m = distance_km * 1000
    return distance_m

In [23]:
converted_distance = kilometers_to_meters(3.7)
print(converted_distance)

3700.0


A variable created inside a function is **local** to that function. In the
example above, `distance_m` exists while the function is running, but the main
notebook receives only the returned value. Keeping temporary names local makes
a long notebook less likely to reuse a variable accidentally.

In [24]:
def celsius_to_fahrenheit(temperature_c):
    return 9 / 5 * temperature_c + 32

print(celsius_to_fahrenheit(20))

68.0


## 11. Importing scientific libraries

Python includes basic language tools, while separately installed **libraries**
provide specialized features. A library is also commonly called a module or
package. The `import` instruction makes a library available in the notebook.

NumPy provides numerical arrays and linear algebra. Matplotlib provides plots.
The short names `np` and `plt` are widely used conventions; they make examples
easier for other Python users to recognize.

In [25]:
import matplotlib.pyplot as plt
import numpy as np

The period in `np.sqrt` means “find the name `sqrt` inside NumPy.” This keeps
related tools together and tells the reader which library supplied the
function.

In [26]:
print(np.pi)
print(np.sqrt(81))

3.141592653589793
9.0


## 12. NumPy arrays

A NumPy **array** is a rectangular collection of values. Unlike a general Python
list, an array normally stores one consistent type and supports efficient
numerical calculations over all its values.

A one-dimensional array can represent one measurement at each station. A
two-dimensional array has rows and columns, like a table. `np.array` converts a
Python list, or a list of lists, into an array.

In [27]:
elevations_m = np.array([1580.0, 1612.0, 1598.0, 1631.0])
print(elevations_m)

[1580. 1612. 1598. 1631.]


In [28]:
station_xy_m = np.array([[0.0, 0.0], [1000.0, 500.0], [1800.0, 1400.0]])
print(station_xy_m)

[[   0.    0.]
 [1000.  500.]
 [1800. 1400.]]


The array's `dtype` reports how its values are stored. Scientific measurements
commonly use floating-point numbers. NumPy arrays are most predictable when
they contain a single type rather than a mixture of numbers and text.

In [29]:
print(elevations_m.dtype)
print(station_xy_m.dtype)

float64
float64


## 13. Shape, axes, indexing, and slicing

An array's **shape** tells us its length along each direction, or **axis**.
`station_xy_m` has shape `(3, 2)`: three rows and two columns. Here each row is a
station, the first column is x position, and the second column is y position.
The order has scientific meaning, so always know what each axis represents.

`ndim` is the number of axes, and `size` is the total number of values.

In [30]:
print(station_xy_m.shape)
print(station_xy_m.ndim, station_xy_m.size)

(3, 2)
2 6


Two indices select one value from a two-dimensional array: row first, then
column. A colon means “all positions along this axis.” Therefore `[:, 0]`
selects every row from column zero.

In [31]:
print(station_xy_m[1, 0])
print(station_xy_m[:, 0])
print(station_xy_m[0, :])

1000.0
[   0. 1000. 1800.]
[0. 0.]


Comparisons can select values too. The expression inside the brackets creates
one `True` or `False` for each elevation, and NumPy keeps only the values whose
comparison is true. This is called Boolean selection.

In [32]:
high_elevations = elevations_m[elevations_m > 1600]
print(high_elevations)

[1612. 1631.]


**Checkpoint.** If an array has shape `(8, 3)`, how many rows, columns, and total
values does it contain?

<details><summary>Answer</summary>

It contains 8 rows, 3 columns, and \(8 	imes 3 = 24\) values.

</details>

## 14. Creating regular sequences and changing shape

Scientific calculations often need evenly spaced coordinates. `np.arange`
creates values separated by a chosen step, while `np.linspace` creates a chosen
number of values including both endpoints. `np.zeros` and `np.ones` create
arrays with a requested shape.

In [33]:
sample_numbers = np.arange(0, 6, 1)
distances_km = np.linspace(0, 10, 6)
print(sample_numbers, distances_km)

[0 1 2 3 4 5] [ 0.  2.  4.  6.  8. 10.]


In [34]:
empty_table = np.zeros((3, 2))
print(empty_table)

[[0. 0.]
 [0. 0.]
 [0. 0.]]


`reshape` changes the arrangement but not the number of values. It is useful
when the new axes have a clear meaning. A `(6,)` array can become `(2, 3)`
because both contain six values. It cannot become `(4, 2)` because that would
require eight values.

In [35]:
grid = sample_numbers.reshape(2, 3)
print(grid)
print(grid.shape)

[[0 1 2]
 [3 4 5]]
(2, 3)


## 15. Array arithmetic and vectorization

Ordinary arithmetic on arrays is **elementwise**: NumPy performs the operation
separately at each position. Adding two same-shaped arrays adds corresponding
values. Multiplying by a single number multiplies every value.

In [36]:
depth_km = np.array([1.5, 3.0, 7.2])
depth_m = depth_km * 1000
print(depth_m)

[1500. 3000. 7200.]


Applying one array expression to all values is called **vectorization**. The
word does not mean that you must already know vector mathematics; it simply
describes asking NumPy to do the repeated numerical work. The result below is
the same as the earlier loop, but it states the calculation more directly.

In [37]:
temperature_c = np.array([5.0, 12.0, 20.0, 27.0])
temperature_f = 9 / 5 * temperature_c + 32
print(temperature_f)

[41.  53.6 68.  80.6]


NumPy also provides summary functions. A **mean** is the arithmetic average,
while a **standard deviation** describes typical spread around that average.

In [38]:
print(np.mean(elevations_m))
print(np.min(elevations_m), np.max(elevations_m))
print(np.std(elevations_m))

1605.25
1580.0 1631.0
18.699933154960743


## 16. Broadcasting: combining compatible shapes

**Broadcasting** is NumPy's rule for combining arrays whose shapes differ in a
compatible way. You already used its simplest form when multiplying an entire
array by one number: NumPy applied that single number to every position.

A column with shape `(3, 1)` can combine with a row of shape `(4,)`. NumPy uses
each of the three column values across all four row positions, producing a
`(3, 4)` table. `np.newaxis` adds the length-one axis that makes the column role
explicit.

In [39]:
slopes = np.array([0.5, 1.0, 2.0])[:, np.newaxis]
x_values = np.array([0.0, 1.0, 2.0, 3.0])
lines = slopes * x_values
print(lines)

[[0.  0.5 1.  1.5]
 [0.  1.  2.  3. ]
 [0.  2.  4.  6. ]]


In [40]:
print(slopes.shape, x_values.shape)
print(lines.shape)

(3, 1) (4,)
(3, 4)


Do not memorize broadcasting as a trick. First decide what each row and column
should mean, then inspect the input and output shapes. A calculation that runs
successfully may still be scientifically wrong if an axis has the wrong
meaning.

## 17. Making a scientific plot

A plot translates numerical values into a visual comparison. Matplotlib calls
the whole drawing a **figure** and the region containing one set of axes an
**Axes object**. The common line `fig, ax = plt.subplots()` creates both.

A scientific plot should identify the quantity and unit on each axis. A title
or legend should explain what is being compared. These labels are part of the
result, not optional decoration.

In [41]:
distance_km = np.linspace(0, 20, 21)
elevation_m = 1600 + 80 * np.sin(distance_km / 4)
fig, ax = plt.subplots(figsize=(6, 3))
ax.plot(distance_km, elevation_m)

In [42]:
ax.set_xlabel("Distance along profile (km)")
ax.set_ylabel("Elevation (m)")
ax.set_title("Example topographic profile")
ax.grid(alpha=0.3)

A scatter plot displays individual observations rather than connecting them.
The semicolon at the end suppresses Jupyter's text representation of the final
plotting object; it does not change the figure.

In [43]:
fig, ax = plt.subplots(figsize=(5, 3))
ax.scatter([1, 2, 3, 4], [2.1, 2.8, 4.2, 4.9])
ax.set(xlabel="Time (days)", ylabel="Displacement (mm)",
       title="Four displacement observations");

## 18. Vectors, matrices, and matrix multiplication

In numerical work, a one-dimensional array is often used as a **vector**, an
ordered sequence of numbers. A two-dimensional array is often used as a
**matrix**, a rectangular table of numbers. These names describe the
mathematical role of the arrays.

Matrix multiplication combines rows of the first matrix with columns of the
second. In Python, `@` means matrix multiplication. The inner dimensions must
match: a matrix with shape `(3, 2)` can multiply a vector with shape `(2,)`,
producing a vector with shape `(3,)`.

In [44]:
matrix = np.array([[1.0, 0.0], [1.0, 1.0], [1.0, 2.0]])
vector = np.array([2.0, 0.5])
prediction = matrix @ vector
print(prediction)

[2.  2.5 3. ]


The `*` symbol still means elementwise multiplication. It does not perform the
row-by-column calculation. For same-shaped matrices, `matrix * matrix` squares
each individual entry, while `matrix @ matrix` may be impossible if the inner
dimensions do not match.

In [45]:
square = np.array([[1.0, 2.0], [3.0, 4.0]])
print(square * square)
print(square @ square)

[[ 1.  4.]
 [ 9. 16.]]
[[ 7. 10.]
 [15. 22.]]


A **transpose** exchanges rows and columns. The `.T` notation returns the
transpose of an array. Transposes appear often when matrix equations combine
observations and unknown values.

In [46]:
print(matrix.shape)
print(matrix.T.shape)
print(matrix.T)

(3, 2)
(2, 3)
[[1. 1. 1.]
 [0. 1. 2.]]


## 19. Understanding matrix inverses

A matrix equation such as \(A x = b\) is a compact way to describe several
linear equations. The matrix \(A\) contains known coefficients, \(b\) contains
known outcomes, and \(x\) contains the unknown values we want to find.

For a suitable square matrix, the **inverse** \(A^{-1}\) reverses the matrix's
effect: \(A^{-1}A\) is the identity matrix, which plays the role of 1 in matrix
multiplication. NumPy can display an inverse with `np.linalg.inv`.

In [47]:
coefficients = np.array([[2.0, 1.0], [1.0, 3.0]])
inverse = np.linalg.inv(coefficients)
print(inverse)

[[ 0.6 -0.2]
 [-0.2  0.4]]


In [48]:
identity_check = inverse @ coefficients
print(identity_check)

[[ 1.00000000e+00 -5.55111512e-17]
 [ 0.00000000e+00  1.00000000e+00]]


## 20. Solving a square system of equations

Although the inverse is useful for understanding the mathematics, computer programs should normally use `np.linalg.solve(A, b)` to solve $A x = b$. `solve` avoids constructing the whole inverse and is generally faster and less affected by numerical roundoff.

In [49]:
observations = np.array([5.0, 7.0])
solution = np.linalg.solve(coefficients, observations)
print(solution)
print(coefficients @ solution)

[1.6 1.8]
[5. 7.]


The last output matches the original observations `[5, 7]`. Substituting a solution back into the original equation is a simple and powerful check: it asks whether the calculated unknowns actually produce the known result.

## 21. Least squares when there are extra observations

Many scientific problems contain more observations than unknown values. No single line may pass exactly through every noisy observation. A **least-squares solution** chooses the model whose predictions have the smallest total squared difference from the data. `np.linalg.lstsq` handles this rectangular case.

In [50]:
design = np.array([[1.0, 0.0], [1.0, 1.0], [1.0, 2.0]])
data = np.array([1.1, 2.0, 3.2])
line_parameters, *_ = np.linalg.lstsq(design, data, rcond=None)
print(line_parameters)

[1.05 1.05]


In [51]:
fitted_data = design @ line_parameters
print(fitted_data)
print(data - fitted_data)

[1.05 2.1  3.15]
[ 0.05 -0.1   0.05]


## 22. Floating-point numbers and reproducible randomness

Computers store most decimal measurements approximately. As a result, a simple
calculation may differ from its exact decimal value by an extremely small
amount. Use `np.isclose` for one calculated number or `np.allclose` for arrays
instead of demanding exact equality with `==`.

In [52]:
calculated = 0.1 + 0.2
print(calculated == 0.3)
print(np.isclose(calculated, 0.3))

False
True


Random numbers are useful for simulating measurement noise. A **seed** is a
starting value that makes a pseudorandom sequence repeatable. Recording the seed
means another student can reproduce the same example exactly.

In [53]:
rng = np.random.default_rng(42)
simulated_noise = rng.normal(loc=0.0, scale=0.5, size=5)
print(simulated_noise)

[ 0.15235854 -0.51999205  0.3752256   0.47028236 -0.97551759]


## 23. Common beginner mistakes and how to recover

Most early programming mistakes come from a few understandable mismatches: the notebook remembers a different value than you expect, an index is one position away from the intended item, or Python and NumPy assign different meanings to similar-looking operations. When a result is surprising, restart and run all cells, print the relevant values and shapes, and reduce the calculation to a tiny example.

In [54]:
python_list = [1, 2, 3]
numpy_array = np.array([1, 2, 3])
print(python_list * 2)
print(numpy_array * 2)

[1, 2, 3, 1, 2, 3]
[2 4 6]


The list is repeated, while every value in the NumPy array is multiplied. Neither behavior is universally better; the correct choice depends on what you intend the values to mean. Similar checks help with `=` versus `==`, zero-based indexing, `*` versus `@`, and omitted plot units. Ask first what the result should represent, then choose the operation.

## 24. Final self-check

Before continuing, try to answer these questions without running code.

1. Why can running notebook cells out of order produce a surprising result?
2. What is the difference between `=` and `==`?
3. Which item does `values[0]` select? Which items does `values[:3]` select?
4. What do the two numbers in an array shape such as `(12, 3)` describe?
5. When does `*` differ from `@`?
6. Why is `np.linalg.solve(A, b)` preferred to `np.linalg.inv(A) @ b`?

<details><summary>Answers</summary>

1. The notebook remembers the most recently run assignments, not merely the
   visible top-to-bottom order.
2. `=` assigns a value to a name; `==` compares two values and returns `True` or
   `False`.
3. Index 0 is the first item. The slice `:3` selects positions 0, 1, and 2.
4. They describe the length of the first and second axes: commonly 12 rows and
   3 columns.
5. For arrays, `*` multiplies corresponding entries. `@` performs matrix
   multiplication by combining rows and columns.
6. `solve` avoids explicitly constructing an inverse, making it generally faster
   and more numerically reliable.

</details>

## Exercises

Work through these in new cells. The matching notebook
`solutions/00_preflight_solutions.ipynb` explains one possible solution to each.

1. **Variables and units.** Store a sample name, a depth of 6.4 km, and a density
   of 2710 kg/m³. Print one readable sentence containing all three values.
2. **Lists and slices.** Make a list containing five field temperatures. Select
   the first two, the last two, and the third value by itself.
3. **A condition.** Classify an earthquake magnitude as `small` below 4,
   `moderate` from 4 up to 6, and `large` at 6 or above.
4. **A loop and a function.** Write a function that converts millimeters to
   meters, then use a loop to apply it to `[2, 5, 12]` mm.
5. **Array selection.** Create `[1.2, 4.8, 3.1, 7.5, 2.0]` as a NumPy array.
   Find its mean and select only values greater than 3.
6. **Vectorization and broadcasting.** Evaluate `slope * x + intercept` for
   slopes 1, 2, and 4; x values from 0 through 5; and intercept 0.5. The result
   should have shape `(3, 6)`.
7. **Plotting.** Plot one of the three lines from Exercise 6. Add a descriptive
   title and label both axes with a quantity and unit.
8. **Matrix multiplication.** Predict three values using
   `A = [[1, 0], [1, 1], [1, 2]]` and `x = [3, 0.4]`. Explain the meaning of
   the output shape.
9. **A linear solve.** Solve `A x = b` for
   `A = [[3, 1], [1, 2]]` and `b = [9, 8]`. Verify the answer by multiplying
   `A @ x`.
10. **Least squares.** Fit a straight line to x values `[0, 1, 2, 3]` and data
    `[1.0, 2.2, 2.9, 4.1]` using `np.linalg.lstsq`. Calculate the predicted data
    and the differences between observations and predictions.

## Recap and next step

You began by running one cell and finished by solving a small matrix problem.
The most important habits are not memorizing every function name. They are
running cells in a known order, choosing meaningful variable names, tracking
units, inspecting array shapes, and checking whether a numerical result makes
sense.

Chapter 01 will introduce GeoDef itself. When unfamiliar Python or NumPy syntax
appears later in the course, return here and experiment with a small version of
the operation.

## Further reading

- [The Python Tutorial](https://docs.python.org/3/tutorial/) provides a broader
  introduction to the language.
- [NumPy: the absolute basics for beginners](https://numpy.org/doc/stable/user/absolute_beginners.html)
  expands on arrays, indexing, and shapes.
- [Matplotlib quick start guide](https://matplotlib.org/stable/users/explain/quick_start.html)
  introduces figures, axes, and common plot types.
- The next course chapter is `01_forward_model.ipynb`.